File: packages/tf-lang-l0-ts/src/vm/interpreter.ts
import type { Program } from '../model/bytecode.js';
import type { Host } from './opcode.js';
import type { Value, World, JournalEntry } from '../model/types.js';
import { canonicalJsonBytes, blake3hex } from '../canon/index.js';
import { emit } from '../proof/dev.js';

export class VM {
  constructor(public host: Host) {}

  private get(regs: Value[], idx: number): Value {
    if (idx < 0 || idx >= regs.length) throw new Error(`register out of bounds: r${idx}`);
    return regs[idx];
  }

  async run(prog: Program): Promise<Value | null> {
    const regs: Value[] = Array.from({ length: prog.regs }, () => null);
    let initialState = structuredClone(regs[0]);
    let initCaptured = false;

    for (const ins of prog.instrs) {
      switch (ins.op) {
        case 'HALT': break;
        case 'CONST': regs[ins.dst] = structuredClone(ins.value); break;
        case 'PACK': {
          const arr = ins.regs.map(r => structuredClone(this.get(regs, r)));
          regs[ins.dst] = { tag: ins.tag, values: arr };
          break;
        }
        case 'UNPACK': {
          const v = this.get(regs, ins.src);
          if (!v || typeof v !== 'object' || v.tag !== ins.tag || !Array.isArray(v.values)) {
            throw new Error('UNPACK expects {tag,values[]} with matching tag');
          }
          if (v.values.length !== ins.dsts.length) throw new Error('UNPACK arity mismatch');
          v.values.forEach((vv: any, i: number) => { regs[ins.dsts[i]] = structuredClone(vv); });
          break;
        }
        case 'ID_HASH': {
          const bytes = canonicalJsonBytes(this.get(regs, ins.src));
          regs[ins.dst] = blake3hex(bytes);
          break;
        }
        case 'SNAP_MAKE': regs[ins.dst] = await this.host.snapshot_make(this.get(regs, ins.state)); break;
        case 'SNAP_ID': regs[ins.dst] = await this.host.snapshot_id(this.get(regs, ins.snapshot)); break;
        case 'LENS_PROJ': {
          regs[ins.dst] = await this.host.lens_project(this.get(regs, ins.state), ins.region);
          emit({ kind: 'Transport', op: 'LENS_PROJ', region: ins.region });
          break;
        }
        case 'LENS_MERGE': {
          regs[ins.dst] = await this.host.lens_merge(this.get(regs, ins.state), ins.region, this.get(regs, ins.sub));
          emit({ kind: 'Transport', op: 'LENS_MERGE', region: ins.region });
          break;
        }
        case 'PLAN_SIM': {
          const res: any = await this.host.call_tf("tf://plan/simulate@0.1", [this.get(regs, ins.world), this.get(regs, ins.plan)]);
          regs[ins.dst_delta] = res?.delta ?? null;
          regs[ins.dst_world] = res?.world ?? null;
          break;
        }
        case 'DIFF_APPLY': regs[ins.dst] = await this.host.diff_apply(this.get(regs, ins.state), this.get(regs, ins.delta)); break;
        case 'DIFF_INVERT': regs[ins.dst] = await this.host.diff_invert(this.get(regs, ins.delta)); break;
        case 'JOURNAL_REC': {
          const s0 = this.get(regs, ins.snap0); if (typeof s0 !== 'string') throw new Error('snap0 not string');
          const s1 = this.get(regs, ins.snap1); if (typeof s1 !== 'string') throw new Error('snap1 not string');
          const je = await this.host.journal_record(this.get(regs, ins.plan), this.get(regs, ins.delta), s0, s1, this.get(regs, ins.meta));
          regs[ins.dst] = je.value;
          break;
        }
        case 'JOURNAL_REW': {
          const w: World = { value: this.get(regs, ins.world) };
          const j: JournalEntry = { value: this.get(regs, ins.entry) };
          const prev = await this.host.journal_rewind(w, j);
          regs[ins.dst] = prev.value;
          break;
        }
        case 'CALL': {
          const args = ins.args.map(a => this.get(regs, a));
          const out = await this.host.call_tf(ins.tf_id, args);
          if (out === null) {
            emit({ kind: 'Conservativity', callee: ins.tf_id, expected: 'value', found: 'null' });
          }
          regs[ins.dst] = out;
          break;
        }
        case 'ASSERT': {
          const v = this.get(regs, ins.pred);
          if (v !== true) {
            emit({ kind: 'Refutation', code: 'E_ASSERT', msg: ins.msg });
            throw new Error(`ASSERT failed: ${ins.msg}`);
          }
          break;
        }
        default: {
          const _: never = ins;
          throw new Error('unknown opcode');
        }
      }
      if (!initCaptured && regs[0] !== null) {
        initialState = structuredClone(regs[0]);
        initCaptured = true;
      }
    }

    const finalState = regs[0];
    // Compute delta locally to mirror Rust VM:
    // identity => null; otherwise full replace
    const a = canonicalJsonBytes(initialState);
    const b = canonicalJsonBytes(finalState);
    const delta = Buffer.from(a).equals(Buffer.from(b)) ? null : { replace: finalState };
    emit({ kind: 'Witness', delta, effect: { read: [], write: [], external: [] } });
    emit({ kind: 'Normalization', target: 'delta' });
    emit({ kind: 'Normalization', target: 'effect' });
    return delta;
  }
}

File: packages/tf-lang-l0-rs/src/vm/interpreter.rs
use crate::canon::{blake3_hex, canonical_json_bytes};
use crate::model::bytecode::Instr;
use crate::model::{JournalEntry, Program, World};
use crate::vm::opcode::Host;
use crate::proof::Replace;
#[cfg(feature = "dev_proofs")]
use crate::proof::{ProofTag, Effect, NormalizationTarget, TransportOp};
#[cfg(feature = "dev_proofs")]
use crate::proof::ProofMeta;
#[cfg(feature = "dev_proofs")]
use std::time::{SystemTime, UNIX_EPOCH};
use serde_json::Value;

/// Simple VM running SSA bytecode with JSON values as registers.
pub struct VM<'h> {
    pub host: &'h dyn Host,
}

#[cfg(feature = "dev_proofs")]
fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

#[cfg(feature = "dev_proofs")]
fn base_meta() -> ProofMeta<'static> {
    ProofMeta {
        runtime: "rust",
        ts: now_ms(),
        region: None,
        gate: None,
        oracle: None,
        seed: None,
    }
}

#[derive(thiserror::Error, Debug)]
pub enum VmError {
    #[error("register out of bounds: r{0} (regs={1})")]
    RegOutOfBounds(u16, usize),
    #[error("invalid bytecode: {0}")]
    Invalid(String),
}

impl<'h> VM<'h> {
    pub fn run(&self, prog: &Program) -> anyhow::Result<Value> {
        let mut regs: Vec<Value> = vec![serde_json::Value::Null; prog.regs as usize];
        let mut initial_state = regs[0].clone();
        let mut init_captured = false;

        // helper to read a register with explicit lifetime
        fn get<'a>(r: u16, regs: &'a Vec<Value>) -> anyhow::Result<&'a Value> {
            match regs.get(r as usize) {
                Some(v) => Ok(v),
                None => Err(VmError::RegOutOfBounds(r, regs.len()).into()),
            }
        }

        for ins in &prog.instrs {
            match ins {
                Instr::Halt => break,
                Instr::Assert { pred, msg } => {
                    let v = get(*pred, &regs)?;
                    if !v.as_bool().unwrap_or(false) {
                        #[cfg(feature = "dev_proofs")]
                        if crate::proof::dev_proofs_enabled() {
                            let mut meta = base_meta();
                            meta.gate = Some("assert".into());
                            crate::emit_tag!(
                                ProofTag::Refutation { code: "ASSERT".into(), msg: Some(msg.clone()) },
                                meta
                            );
                        }
                        return Err(VmError::Invalid(format!("ASSERT failed: {}", msg)).into());
                    }
                }
                Instr::Const { dst, value } => {
                    regs[*dst as usize] = value.clone();
                }
                Instr::Pack {
                    dst,
                    tag,
                    regs: rgs,
                } => {
                    let mut arr = Vec::with_capacity(rgs.len());
                    for r in rgs {
                        arr.push(get(*r, &regs)?.clone());
                    }
                    let obj = serde_json::json!({ "tag": tag, "values": arr });
                    regs[*dst as usize] = obj;
                }
                Instr::Unpack { dsts, tag, src } => {
                    let v = get(*src, &regs)?;
                    let o = v
                        .as_object()
                        .ok_or_else(|| VmError::Invalid("UNPACK expects object".into()))?;
                    let found_tag = o
                        .get("tag")
                        .and_then(|t| t.as_str())
                        .ok_or_else(|| VmError::Invalid("UNPACK missing tag".into()))?;
                    if found_tag != tag {
                        return Err(VmError::Invalid(format!(
                            "UNPACK tag mismatch: expected {}, got {}",
                            tag, found_tag
                        ))
                        .into());
                    }
                    let values = o
                        .get("values")
                        .and_then(|v| v.as_array())
                        .ok_or_else(|| VmError::Invalid("UNPACK missing values".into()))?
                        .clone();
                    if values.len() != dsts.len() {
                        return Err(VmError::Invalid("UNPACK arity mismatch".into()).into());
                    }
                    for (i, d) in dsts.iter().enumerate() {
                        regs[*d as usize] = values[i].clone();
                    }
                }
                Instr::IdHash { dst, src } => {
                    let bytes = canonical_json_bytes(get(*src, &regs)?)?;
                    regs[*dst as usize] = serde_json::json!(blake3_hex(&bytes));
                }
                Instr::SnapMake { dst, state } => {
                    let snap = self.host.snapshot_make(get(*state, &regs)?)?;
                    regs[*dst as usize] = snap;
                }
                Instr::SnapId { dst, snapshot } => {
                    let id = self.host.snapshot_id(get(*snapshot, &regs)?)?;
                    regs[*dst as usize] = serde_json::json!(id);
                }
                Instr::LensProj { dst, state, region } => {
                    let sub = self.host.lens_project(get(*state, &regs)?, region)?;
                    #[cfg(feature = "dev_proofs")]
                    if crate::proof::dev_proofs_enabled() {
                        let mut meta = base_meta();
                        meta.region = Some(region.clone());
                        meta.gate = Some("lens_project".into());
                        crate::emit_tag!(
                            ProofTag::Transport { op: TransportOp::LensProj, region: region.clone() },
                            meta
                        );
                    }
                    regs[*dst as usize] = sub;
                }
                Instr::LensMerge {
                    dst,
                    state,
                    region,
                    sub,
                } => {
                    let merged = self
                        .host
                        .lens_merge(get(*state, &regs)?, region, get(*sub, &regs)?)?;
                    #[cfg(feature = "dev_proofs")]
                    if crate::proof::dev_proofs_enabled() {
                        let mut meta = base_meta();
                        meta.region = Some(region.clone());
                        meta.gate = Some("lens_merge".into());
                        crate::emit_tag!(
                            ProofTag::Transport { op: TransportOp::LensMerge, region: region.clone() },
                            meta
                        );
                    }
                    regs[*dst as usize] = merged;
                }
                Instr::PlanSim {
                    dst_delta,
                    dst_world,
                    world,
                    plan,
                } => {
                    // A placeholder: encode the pair (Δ, W') via a TF call
                    let res = self.host.call_tf(
                        "tf://plan/simulate@0.1",
                        &[get(*world, &regs)?.clone(), get(*plan, &regs)?.clone()],
                    )?;
                    let obj = res.as_object().ok_or_else(|| {
                        VmError::Invalid("PlanSim expects object {delta, world}".into())
                    })?;
                    regs[*dst_delta as usize] =
                        obj.get("delta").cloned().unwrap_or(serde_json::Value::Null);
                    regs[*dst_world as usize] =
                        obj.get("world").cloned().unwrap_or(serde_json::Value::Null);
                }
                Instr::DiffApply { dst, state, delta } => {
                    let out = self
                        .host
                        .diff_apply(get(*state, &regs)?, get(*delta, &regs)?)?;
                    regs[*dst as usize] = out;
                }
                Instr::DiffInvert { dst, delta } => {
                    let inv = self.host.diff_invert(get(*delta, &regs)?)?;
                    regs[*dst as usize] = inv;
                }
                Instr::JournalRec {
                    dst,
                    plan,
                    delta,
                    snap0,
                    snap1,
                    meta,
                } => {
                    let s0 = get(*snap0, &regs)?
                        .as_str()
                        .ok_or_else(|| VmError::Invalid("snap0 not string".into()))?;
                    let s1 = get(*snap1, &regs)?
                        .as_str()
                        .ok_or_else(|| VmError::Invalid("snap1 not string".into()))?;
                    let j = self.host.journal_record(
                        get(*plan, &regs)?,
                        get(*delta, &regs)?,
                        s0,
                        s1,
                        get(*meta, &regs)?,
                    )?;
                    regs[*dst as usize] = j.0;
                }
                Instr::JournalRew { dst, world, entry } => {
                    let w = World(get(*world, &regs)?.clone());
                    let j = JournalEntry(get(*entry, &regs)?.clone());
                    let wprev = self.host.journal_rewind(&w, &j)?;
                    regs[*dst as usize] = wprev.0;
                }
                Instr::Call { dst, tf_id, args } => {
                    let mut a = Vec::with_capacity(args.len());
                    for r in args {
                        a.push(get(*r, &regs)?.clone());
                    }
                    let out = self.host.call_tf(tf_id, &a).map_err(|e| {
                        #[cfg(feature = "dev_proofs")]
                        if crate::proof::dev_proofs_enabled() {
                            let mut meta = base_meta();
                            meta.gate = Some("call_tf".into());
                            meta.oracle = Some(tf_id.clone());
                            crate::emit_tag!(
                                ProofTag::Conservativity {
                                    callee: tf_id.clone(),
                                    expected: "ok".into(),
                                    found: format!("{}", e)
                                },
                                meta
                            );
                        }
                        e
                    })?;
                    regs[*dst as usize] = out;
                }
            }
            if !init_captured && regs[0] != serde_json::Value::Null {
                initial_state = regs[0].clone();
                init_captured = true;
            }
        }

        let final_state = regs.get(0).cloned().unwrap_or(serde_json::Value::Null);
        let delta = if final_state == initial_state {
            None
        } else {
            Some(Replace { replace: final_state.clone() })
        };
        #[cfg(feature = "dev_proofs")]
        if crate::proof::dev_proofs_enabled() {
            let mut witness_meta = base_meta();
            witness_meta.gate = Some("vm.run".into());
            crate::emit_tag!(
                ProofTag::Witness { delta: delta.clone(), effect: Effect::default() },
                witness_meta
            );
            for target in [NormalizationTarget::Delta, NormalizationTarget::Effect] {
                let mut meta = base_meta();
                meta.gate = Some("vm.run".into());
                crate::emit_tag!(ProofTag::Normalization { target }, meta);
            }
        }
        let out = match delta {
            None => serde_json::Value::Null,
            Some(d) => serde_json::json!({ "replace": d.replace }),
        };

        Ok(out)
    }
}
